src/cmd/serve.rs

```diff
@@ -690,52 +704,85 @@ pub fn serve(
-                            continue;
-                        }
+                    // We currently only handle notify events that report a single path per event.
+                    // Enforce that assumption to avoid surprises. See `get_relevant_event_kind`.
+                    assert!(
+                        event.event.paths.len() == 1,
+                        "Didn't expect notify event with multiple paths."
+                    );
+                    let path = event.event.paths[0].clone();
 
-                        if is_temp_file(&path) {
-                            continue;
-                        }
+                    if is_ignored_file(&site.config.ignored_content_globset, &path) {
+                        continue;
+                    }
 
-                        // We only care about changes in non-empty folders
-                        if path.is_dir() && is_folder_empty(&path) {
-                            continue;
-                        }
+                    if is_temp_file(&path) {
+                        continue;
+                    }
+
+                    // We only care about changes in non-empty folders
+                    if path.is_dir() && is_folder_empty(&path) {
+                        continue;
+                    }
+
+                    meaningful_events.push((path, simple_kind.unwrap(), event.time));
+                }
+
+                if meaningful_events.is_empty() {
+                    continue;
+                }
 
-                        let format =
-                            format_description!("[year]-[month]-[day] [hour]:[minute]:[second]");
-                        let current_time =
-                            OffsetDateTime::now_utc().to_offset(utc_offset).format(&format);
-                        if let Ok(time_str) = current_time {
-                            println!("Change detected @ {}", time_str);
-                        } else {
-                            // if formatting fails for some reason
-                            println!("Change detected");
-                        };
-
-                        let start = Instant::now();
-                        match detect_change_kind(root_dir, &path, &config_path) {
-                            (ChangeKind::Content, _) => {
-                                console::info(&format!("-> Content changed {}", path.display()));
+                // Editors such as Vim will create numerous events when operating on a single file.
+                // These events are not combined above. Take the latest event for each unique path.
+                let filtered_events = debounce_events(&meaningful_events);
+
+                // Map of change_kind -> (partial_path, full_path, event_kind).
+                let changes: HashMap<ChangeKind, Vec<(PathBuf, &PathBuf, &SimpleFSEventKind)>> =
+                    map_events_to_change_map(&filtered_events, &root_dir, &config_path);
```

Comment:
Can we extract the whole logic of finding and sorting/filtering events to a separate function so it's easy to test and doesn't take space in the actual loop?

Reply:
Thanks for your time. Before I start, to confirm, are you asking about moving only the new code in this merge request or also existing code like `is_ignored_file`, `is_temp_file`, etc. participating in the filtering?

Reply:
I think it might make sense to move everything to a new file? But no strong feeling about it

Reply:
Thanks. In that case, I will focus on just my changes for now.

Reply:
Done.